### DISCUSSION TITLE
Changing FieldArray keyName and using reset in a child component causes infinity renders

### DISCUSSION THREAD

**Discussion by TonyVeigel** [react-hook-form/react-hook-form#5883] [Q&A] [answered]
**Describe the bug**
Changing FieldArray keyName and using it as the key (key={field.keyId}) causes infinity renders when using reset. Are you not supposed to use the updated keyName as the key? 

**To Reproduce**
I'm using reset in the nestedFieldArray with a keyName change on the field array.
https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-forked-ddurk?file=/src/fieldArray.js

**Comment by bluebill1049** [comment]
it should worked just fine: https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-forked-z2ewl?file=/src/nestedFieldArray.js

**Reply by TonyVeigel** [reply]
You're not resetting the form anywhere. Call reset from useEffect in nestedFieldArray like in my example.

**Reply by bluebill1049** [reply]
updated.

**Reply by bluebill1049** [reply]
i see your problem, you are creating a loop there... you are reset inside the nested field array, which re-render the entire field array again and again...

**Reply by TonyVeigel** [reply]
That shouldn't be a problem though. Why is reset causing that to happen? It shouldn't be causing the component to unmount and remount because I reset something.

**Reply by TonyVeigel** [reply]
Do you not consider this an issue? In our forms we grab additional async data in a components for drop downs. Since reset for some reason causes remounting (which it shouldn't), it causes all the drop downs to refetch data every time someone resets a form to its original state.

**Comment by thanh-nguyen-95** [comment, marked as answer]
I don't think changing `keyName` in `useFieldArray` is related to the issue you're facing. A render loop is created with how the components are structured. It looks something like this:

**General structure**

- `App` --> `Fields` --> `NestedFields`

**Render flow**

- `App` -- _reset(defaultValue)_ --> `Fields` re-renders because fields changed --> `NestedFields` re-renders because its parent re-renders -- _reset()_ gets called, which resets everything --> `Fields` re-renders because fields changed --> `NestedFields` re-renders once more because its parent re-renders -- _reset()_ gets called again --> rinse and repeat

Removing `reset()` inside `nestedArray.jsx` should solve the infinite loop problem. That's as far as we can help you with the current CSB you provided.

**Reply by TonyVeigel** [reply]
Re-render is NOT the same as re-mounting. useEffect((), []) gets called once on mounting, never again unless it is remounted. Re-rendering has nothing to do with it. Re-rendering !== re-mounting.

Example: https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-forked-p1i43?file=/src/fieldArray.js

No matter how you slice it, calling reset should not cause a component to remount just like setting state does not cause a component to re-mount. It is a bug and if you want to admit it or not is a different issue.

**Reply by TonyVeigel** [reply]
@thanh-nguyen-95 @bluebill1049 and the reason it is happening, I just figured out, is because you are using a different UUID for fields every time on reset which causes all the components to be remounted. That is definitely a bug. You should be using the same UUID on reset.

**Reply by thanh-nguyen-95** [reply]
> 
> 
> @thanh-nguyen-95 @bluebill1049 and the reason it is happening, I just figured out, is because you are using a different UUID for fields every time on reset which causes all the components to be remounted. That is definitely a bug. You should be using the same UUID on reset.

https://codesandbox.io/s/infallible-bose-1vi5c?file=/src/child.jsx

I created a CSB to mimic Tony's case and I can confirm that this is a key generation issue (not because of changing `keyName` option as mentioned by @TonyVeigel)

However, using the same key when reset doesn't make sense because in theory you're replace old data with new data so a new key is needed for React to know the data was changed. Keeping the key between renders might be a better call here.

Ps. It is also worth mentioning that re-render and re-mount are 2 different phases according to this diagram https://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/ I was wrong when I use `re-render` instead of `re-mount`. My sincere apology to @TonyVeigel

**Reply by bluebill1049** [reply]
I have my full respect for your kindness and patient @thanh-nguyen-95 

Going back to the original reply, it's creating a loop when you have reset the entire form within the child node.

`mount -> reset -> re-render -> unmount (child nodes) -> remount (child nodes)`

If you are `reset` the form, it doesn't make sense to me to keep the same `key`, pay attention `reset` API is for the entire form, not `update`.
